pkg/visitors/createfv.go

```diff
@@ -0,0 +1,240 @@
+	if bp.Offset < offset {
+		hbp, err := uefi.NewBIOSPadding(bpBuf[:offset-bp.Offset], bp.Offset)
+		if err != nil {
+			return err
+		}
+		newElements = append(newElements, uefi.MakeTyped(hbp))
+	}
+
+	// Add the FV
+	newElements = append(newElements, uefi.MakeTyped(fv))
+
+	// keep part of the BIOS Pad after the new fv if needed
+	if offset-bp.Offset+fv.Length < uint64(len(bpBuf)) {
+		tbp, err := uefi.NewBIOSPadding(bpBuf[offset-bp.Offset+fv.Length:], offset+fv.Length)
+		if err != nil {
+			return err
+		}
+		newElements = append(newElements, uefi.MakeTyped(tbp))
+	}
+
+	// Keep the remaining Elements in the BIOS Region
+	if idx+1 < len(br.Elements) {
+		newElements = append(newElements, br.Elements[idx+1:]...)
+	}
+	br.Elements = newElements
+	return nil
+}
+
+func createEmptyFirmwareVolume(fvOffset, size uint64, name *guid.GUID) (*uefi.FirmwareVolume, error) {
+	// TODO: can this be refactored with the code in repack.go and assemble.go ?
+	fv := &uefi.FirmwareVolume{} // new Firmware Volume
+	// Set up volume header first.
+	fv.FileSystemGUID = *uefi.FFS2
+	fv.Signature = binary.LittleEndian.Uint32([]byte("_FVH"))
+	// TODO: retrieve all details from (all) other fv in BIOS Region
+	fv.Attributes = 0x0004FEFF
+	fv.Revision = 2
+	// Create Blocks
+	fv.Blocks = make([]uefi.Block, 2)
+	fv.Blocks[0] = uefi.Block{Size: 4096, Count: uint32(size / 4096)}
+	fv.Blocks[1] = uefi.Block{}
+	// Calculate the HeaderLen field
+	fv.HeaderLen = uint16(uefi.FirmwareVolumeFixedHeaderSize + int(unsafe.Sizeof(uefi.Block{}))*len(fv.Blocks))
+
+	fv.DataOffset = uint64(fv.HeaderLen) // unless we add the extended header
+	fv.Length = size
+
+	if name != nil {
+		// TODO: should the extended header offset be computed ?
+		fv.ExtHeaderOffset = 0x60
+		fv.FVName = *name
+		fv.ExtHeaderSize = uefi.FirmwareVolumeExtHeaderMinSize
+	}
+	// Generate binary header.
+	header := new(bytes.Buffer)
+	err := binary.Write(header, binary.LittleEndian, fv.FirmwareVolumeFixedHeader)
+	if err != nil {
+		return nil, fmt.Errorf("unable to construct binary header of new firmware volume: got %v", err)
+	}
+	for _, b := range fv.Blocks {
+		err = binary.Write(header, binary.LittleEndian, b)
+		if err != nil {
+			return nil, fmt.Errorf("unable to construct binary header of new firmware volume: got %v", err)
+		}
+
+	}
+	buf := header.Bytes()
+
+	// Checksum the header
+	sum, err := uefi.Checksum16(buf[:fv.HeaderLen])
+	if err != nil {
+		return nil, err
+	}
+	newSum := 0 - sum
+	binary.LittleEndian.PutUint16(buf[50:], newSum)
+
+	// Store the header buffer in
+	fv.SetBuf(buf)
+
+	if name != nil {
+		// Build the ExtHeader
+		extHeader := new(bytes.Buffer)
+		err = binary.Write(extHeader, binary.LittleEndian, fv.FirmwareVolumeExtHeader)
+		if err != nil {
+			return nil, fmt.Errorf("unable to construct binary extended header of new firmware volume: got %v", err)
+		}
+
+		extHeaderFile, err := uefi.CreatePadFile(uint64(uefi.FileHeaderMinLength + fv.ExtHeaderSize))
```

Comment:
Is the extended header usually an actual file in the FV? I did not realize that. Can you point to some documentation where that's laid out? thanks.

Reply:
I found that out by the way it's created in tiano

Reply:
https://github.com/tianocore/edk2/blob/master/BaseTools/Source/C/GenFv/GenFvInternalLib.c#L563 the `AddPadFile` func takes the extended header as argument and is used in `GenerateFvImage` here:
https://github.com/tianocore/edk2/blob/master/BaseTools/Source/C/GenFv/GenFvInternalLib.c#L2772

Reply:
I couldn't find a reference in the uefi PI spec.

Reply:
In the flash image there where this 16 0xff's plus 8 bytes with some values before the actual ext-header content ... so I wondered where they came from ... and found the above mentioned code. The content matched with the Pad File header so that was it.

Reply:
I see. That's terrible design. Why did they do that -_-. Do you mind explaining it in comments intthe code so when we come back in the future we won't forget? Thanks

Reply:
Just this last one then I think we're good to go